src/util.c

```diff
@@ -40,9 +40,11 @@
 #include <stdint.h>
 #include <errno.h>
 #include <time.h>
+#include <fcntl.h>
 
 #include "util.h"
 #include "sha256.h"
+#include "anet.h"
```

Comment:
I'm not sure if util.c is the best file to put `createPipe()`. Maybe `anet.c` is a better place. So far, util is mostly string manipulation functions.

The commet in anet.c says "Basic TCP socket stuff made a bit less borin", but some functions like `anetNonBlock()` is for any file descriptor, also pipes, so I think createPipe() can also be in anet.

I will leave it for other reviewers to think about.

Reply:
I've thought about putting `createPipe()` into `anet.c` and then ceased for the same reason of you.

Reply:
If we want to keep anet for only TCP, we can avoid using `anetNonBlock()` in `createPipe()`. It's actually nothing else than `fcntl(fd, F_SETFL, O_NONBLOCK)` so I think we can just call fcntl directly here and not include anet.h.

Reply:
Actually `anetNonBlock()` is also called on pipes way before this PR, so I don't think `createPipe()` using `anetNonBlock()` is our main problem here.

Reply:
No, the problem is that util now depends on anet. It is better if they are independent.

I just realized anetNonBlock does one fcntl to read the flags and then another to set the flags. anetCloexec also does two. If we set both flags (non-linux non-freebsd) we do 4 syscalls instead of 1 (fcntl with all flags combined) for one fd.

Actually, I think createPipe() can be much simpler if it's actually setting all flags directly (error handling omitted):

```C
#if defined(__linux__) || defined(__FreeBSD__)
    int flags = read_flags & write_flags;
    pipe2(fds, flags);
    if (read_flags != flags) fcntl(fds[0], read_flags);
    if (write_flags != flags) fcntl(fds[1], write_flags);
#else
    pipe(fds)
    fcntl(fds[0], read_flags);
    fcntl(fds[1], write_flags);
#endif
```

Reply:
`createPipe()` would be concise if we just call `fcntl()` directly instead of `anetNonBlock()` and `anetCloexec()` like you said, what do you think about that?
@oranagra

Reply:
> No, the problem is that util now depends on anet. It is better if they are independent.
> 
> I just realized anetNonBlock does one fcntl to read the flags and then another to set the flags. anetCloexec also does two. If we set both flags (non-linux non-freebsd) we do 4 syscalls instead of 1 (fcntl with all flags combined) for one fd.
> 
> Actually, I think createPipe() can be much simpler if it's actually setting all flags directly (error handling omitted):
> 
> ```c
> #if defined(__linux__) || defined(__FreeBSD__)
>     int flags = read_flags & write_flags;
>     pipe2(fds, flags);
>     if (read_flags != flags) fcntl(fds[0], F_SETFD, read_flags);
>     if (write_flags != flags) fcntl(fds[1], F_SETFD, write_flags);
> #else
>     pipe(fds)
>     fcntl(fds[0], F_SETFD, read_flags);
>     fcntl(fds[1], F_SETFD, write_flags);
> #endif
> ```

I just recall that we can't simply refine the `createPipe()` like that cuz we have to differentiate between `F_GETFL/F_SETFL` and `F_GETFD/F_SETFD`

Reply:
Oh, good point. Yhey're called "file descriptor flags" (FD) and "file status flags" (FL) in the fcntl map page.

I still think we should use fcntl directly instead of the anet functions. Why?

* We avoid the extra syscalls with F_GETFL and F_GETFD.
* anetNonBlock and anetCloexec are not abstracting anything anyway because we know we're working with file descriptors since we call pipe() and pipe2() and we use the constants defined in fcntl.h (O_CLOEXEC and O_NONBLOCK). It's a useless abstraction which only hides what's happening.

Reply:
i'm sorry, i'm not following this discussion closely (busy elsewhere).
regarding this specific thread (haven't read all of it), here's my view.
I think anet.c isn't currently specifically related to TCP, it's a file that contains some platform abstraction utilities (mainly revolved around sockets, and file descriptors), so this pipe related wrapper seems to be a good candidate to be placed there.
on the other hand, util.c is not about platform abstraction, but rather just some pure C code utilities for small chores.

Reply:
OK, I'm moving the `createPipe()` from `util.c` to `anet.c`.